tvm/src/schedule/schedule_dataflow_rewrite.cc

```diff
@@ -304,87 +203,126 @@ void Schedule::to_stage(const Tensor& target,
-  if (is_placeholder) {
-    for (size_t i = 0; i < num_stage; i++) {
-      Stage s = (*this)->stages[i];
-      if (const ExternOpNode* op = s->op.as<ExternOpNode>()) {
-        for (size_t j = 0; j < op->inputs.size(); j++) {
-          if (target == op->inputs[j]) {
-            target_buffer = op->input_placeholders[j];
-            consumers.push_back(s);
-          }
-        }
-      }
-    }
-  } else {  // mark device scope of consumers & update kernel stmts
-    const ExternOpNode* op = target_stage->op.as<ExternOpNode>();
-    target_buffer = op->output_placeholders[0];
-    consumers.push_back(target_stage);
-    for (size_t i = 0; i < num_stage; i++) {
-      Stage s = (*this)->stages[i];
-      if (const ExternOpNode* op = s->op.as<ExternOpNode>()) {
-        for (size_t j = 0; j < op->inputs.size(); j++) {
-          if (target_buffer == op->input_placeholders[j]) {
-            consumers.push_back(s);  // mark buffer in calls
-          }
+  // Extract target buffer and consumers of the channel
+  // When a global buffer is streamed between modules,
+  // it can and only can have two consumers
+  const ExternOpNode* op = target_stage->op.as<ExternOpNode>();
+  if (op == NULL) {
+    LOG(CLEAN) << "Target tensor " << target << " is placeholder "
+               << "and cannot be streamed to an on-chip consumer...";
+    LOG(CLEAN) << "Consider using s.to(" << target->op->name
+               << ", target.xcel)...";
+    return;
+  }
+
+  Buffer target_buffer = op->output_placeholders[0];
+  consumers.push_back(target_stage);
+  for (size_t i = 0; i < num_stage; i++) {
+    Stage s = (*this)->stages[i];
+    if (const ExternOpNode* op = s->op.as<ExternOpNode>()) {
+      for (size_t j = 0; j < op->inputs.size(); j++) {
+        if (target_buffer == op->input_placeholders[j]) {
+          consumers.push_back(s);
         }
       }
     }
   }
 
-  // inter-stage data movement
+  // Inter-stage data movement
   if (stream_pos.size() == 0) {
+    // Self loop-back case
     if (destOp == srcOp) {
-      // mutate loop body (attr_value indicates self-loop)
+      // Annotate the allocate node
+      HCL_DEBUG_LEVEL(2) << "[ debug ] Streaming tensor " << target_buffer
+                         << " to stage " << destOp->name << " (loopback)...";
       VarExpr node(target_buffer->data.node_);
-      Stmt dest_body = AttrStmt::make(node, attr::device_scope,
-                                      IntImm::make(Int(32), 0), destOp->body);
+      Stmt dest_body =
+          AttrStmt::make(node, attr::device_scope,
+                         IntImm::make(Int(32), channel_depth), destOp->body);
       dest->op = ExternOpNode::make(destOp->name, destOp->tag, destOp->axis,
                                     destOp->inputs, destOp->input_placeholders,
                                     destOp->output_placeholders, dest_body);
+      // Stage-to-stage channel.
+      // 1. one-to-one streaming
+      // 2. one-to-many streaming
```

Comment:
Is one-to-many handled by multiple one-to-one?

Reply:
these are removed.